### PR TITLE
Add ops for demo

### PR DESCRIPTION
**Pull request by thecharlieblake**
Just a mini pr to add in some core ops needed for the demo. There will be a larger one at some point with the magic auto-visualisation and unit scaling wrappers, but wanted to get this small thing in before experiments start...

### CODE REVIEW

File: unit_scaling/tests/test_functional.py [hunk 1]
```diff
@@ -282,23 +279,51 @@ def test_linear_scale_for_grad_input() -> None:
...
 # --- test residual() ---
 
 
 def test_residual() -> None:
     for tau in [0.2, 0.5, 0.8]:
-        input = unit_normal(2**10)
+        input = randn(2**10, requires_grad=True)
         residual, skip = residual_split(input, tau)
-        residual = linear(residual, unit_normal(2**10, 2**10), bias=None)
+        residual = linear(residual, randn(2**10, 2**10), bias=None)
```
Comment (DouglasOrr) [inline unit_scaling/tests/test_functional.py:326]: No `requires_grad=True`; does that mean that this grad is not tested?

File: unit_scaling/tests/test_functional.py [hunk 2]
```diff
@@ -282,23 +279,51 @@ def test_linear_scale_for_grad_input() -> None:
...
+# --- test add() ---
+
+
+def test_add() -> None:
+    left = randn(2**8, 2**10, requires_grad=True)
+    right = randn(2**8, 2**10, requires_grad=True)
+    output = add(left, right)
+    unit_backward(output)
+
+    assert_unit_scaled(output, left, right)
```
Comment (DouglasOrr) [inline unit_scaling/tests/test_functional.py:300]: Should this be `left.grad, right.grad`?

File: unit_scaling/tests/test_functional.py [hunk 3]
```diff
@@ -282,23 +279,51 @@ def test_linear_scale_for_grad_input() -> None:
...
+
+
+def test_add_broadcast() -> None:
+    left = randn(2**10, requires_grad=True)
+    right = randn(2**8, 2**10, requires_grad=True)
+    output = add(left, right)
+    unit_backward(output)
+
+    assert_unit_scaled(output, left, right)
```
Comment (DouglasOrr) [inline unit_scaling/tests/test_functional.py:309]: Ditto.

File: unit_scaling/tests/test_functional.py [hunk 4]
```diff
@@ -282,23 +279,51 @@ def test_linear_scale_for_grad_input() -> None:
...
+
+    left = randn(2**8, 1, 2**10, requires_grad=True)
+    right = randn(2**9, 2**10, requires_grad=True)
+    output = add(left, right)
+    unit_backward(output)
+
+    assert_unit_scaled(output, left, right)
```
Comment (DouglasOrr) [inline unit_scaling/tests/test_functional.py:316]: Ditto.

File: unit_scaling/functional.py [hunk 5]
```diff
@@ -52,6 +52,15 @@ def scaled_f(input: Tensor, *args: Any, **kwargs: Any) -> Tensor:
     return scaled_f
 
 
+def _get_broadcast_sizes(*args: Tensor) -> Iterable[int]:
```
Comment (DouglasOrr) [inline unit_scaling/functional.py:55]: Extremely minor; perhaps `Tuple[int, ...]` is a bit more natural here.

File: unit_scaling/functional.py [hunk 6]
```diff
@@ -186,6 +195,21 @@ def layer_norm(
...
+    unsupported_args=["alpha"],
+)
+def add(
+    input: Tensor, other: Tensor, alpha: int = 1, out: Optional[Tensor] = None
+) -> Tensor:
+    input_broadcast_size, other_broadcast_size = _get_broadcast_sizes(input, other)
+    input = scale_bwd(input, input_broadcast_size**-0.5)
+    other = scale_bwd(other, other_broadcast_size**-0.5)
+    out = torch.add(input, other, out=out)
+    return scale_fwd(out, 2**-0.5)
```
Comment (DouglasOrr) [inline unit_scaling/functional.py:210]: Maybe could just support `alpha`, just by forwarding it to `torch.add` & then `scale_fwd(out, (1+alpha)**-0.5)`? Or is this perhaps a bit risky / an anti-pattern?

File: unit_scaling/functional.py [hunk 7]
```diff
@@ -186,6 +195,21 @@ def layer_norm(
     return F.layer_norm(input, normalized_shape, weight, bias, eps)
 
 
+@docstring_from(
+    torch.add,
+    short_description="Applies a **unit-scaled** addition.",
+    unsupported_args=["alpha"],
+)
+def add(
+    input: Tensor, other: Tensor, alpha: int = 1, out: Optional[Tensor] = None
```
Comment (DouglasOrr) [inline unit_scaling/functional.py:204]: Worth adding a `constraint`? If so, would this be the first default=unconstrained op, though?

File: unit_scaling/functional.py [hunk 8]
```diff
@@ -252,6 +276,47 @@ def embedding(
...
+            "probability greater than 0.0 is specified."
+        ),
+    )
+    def scaled_dot_product_attention(
+        query: Tensor,
+        key: Tensor,
+        value: Tensor,
+        attn_mask: Optional[Tensor] = None,
+        dropout_p: float = 0.0,
+        is_causal: bool = False,
```
Comment (DouglasOrr) [inline unit_scaling/functional.py:299]: Should the docs include a warning that unit-scaling is only approximate when using `attn_mask`, `dropout_p` and `is_causal`?

File: unit_scaling/functional.py [hunk 9]
```diff
@@ -252,6 +276,47 @@ def embedding(
...
+        scale: Optional[float] = None,
```
Comment (DouglasOrr) [inline unit_scaling/functional.py:300]: What is the unused `scale` arg for? I can't see it in `torch.scaled_dot_product_attention`.

**Review by DouglasOrr — COMMENTED** [review]
Sorry, this is an annoying number of comments. Feel free to ignore any that aren't helpful.

It looks grand, though. I like the "vanillaisation" of test RNG!

File: unit_scaling/functional.py (hunk 6 above)
Comment (thecharlieblake) [reply unit_scaling/functional.py:210]: Yeah I thought a bit about whether I should push scaling factors through to alpha (and compensate accordingly) - in the end I decided it could be left for now and therefore would be left for now - if you think it should be added at this stage though please push back!

File: unit_scaling/functional.py (hunk 8 above)
Comment (thecharlieblake) [reply unit_scaling/functional.py:299]: I'm tempted to say that given there's a pretty wide range of cases where values only come out approximately unit-scaled, a warning here would mean lots of warnings - for instance, probably the largest thing that pushes us from true US is the correlation introduced by this op. But I'd rather not go into those details at this stage - at some point would be nice to address this more generally in the user guide?

File: unit_scaling/tests/test_functional.py (hunk 1 above)
Comment (thecharlieblake) [reply unit_scaling/tests/test_functional.py:326]: Yeah, this linear is just to simulate having "some kind of op" on the residual branch, so checking the grads of this op's weights is probably out of scope for this test?

File: unit_scaling/functional.py (hunk 6 above)
Comment (DouglasOrr) [reply unit_scaling/functional.py:210]: No objection!

File: unit_scaling/functional.py (hunk 8 above)
Comment (DouglasOrr) [reply unit_scaling/functional.py:299]: Yes, agree that this approach sounds fair.

File: unit_scaling/functional.py [hunk 10]
```diff
@@ -201,13 +201,27 @@ def layer_norm(
...
+    alpha: int = 1,
+    out: Optional[Tensor] = None,
 ) -> Tensor:
     input_broadcast_size, other_broadcast_size = _get_broadcast_sizes(input, other)
-    input = scale_bwd(input, input_broadcast_size**-0.5)
-    other = scale_bwd(other, other_broadcast_size**-0.5)
+    grad_input_scale = input_broadcast_size**-0.5
+    grad_other_scale = other_broadcast_size**-0.5
+    fwd_scale = 2**-0.5
+    if constraint:
```
Comment (thecharlieblake) [inline unit_scaling/functional.py:214]: I have to use this slightly nasty construct (used elsewhere) to handle the different kinds of constraints one is allowed to pass in

File: unit_scaling/tests/test_functional.py (hunk 1 above)
Comment (DouglasOrr) [reply unit_scaling/tests/test_functional.py:337]: Yeah, I guess I was just wondering if it helped demonstrate that the branch gradients themselves are unit scale (vs immediately scaling the backward grads at the beginning of the backward pass.) But no huge need for this.

File: unit_scaling/functional.py (hunk 10 above)
Comment (DouglasOrr) [reply unit_scaling/functional.py:214]: Ah, yes, I remember wondering if this boilerplate could be refactored into an `apply_constraint(constraint, ...)` somehow.

**Review by DouglasOrr — APPROVED** [review]
